### DISCUSSION TITLE
Leaf API fails to start

### DISCUSSION THREAD

**Discussion by artgoldberg** [uwrit/leaf#410] [Q&A]
Hi Folks

I'm following the [Leaf installation directions](https://leafdocs.rit.uw.edu/installation/). In step 8a `systemctl start leaf_api.service` fails to start the Leaf API. This is the error reported by `systemctl status`:

    ● leaf_api.service - Leaf API Service
       Loaded: loaded (/var/opt/leafapi/services/leaf_api.service; linked; vendor preset: disabled)
       Active: failed (Result: signal) since Sat 2021-05-08 12:11:33 EDT; 1min 0s ago
      Process: 90214 ExecStart=/usr/bin/dotnet API.dll (code=killed, signal=ABRT)
     Main PID: 90214 (code=killed, signal=ABRT)

    May 08 12:11:33 omop-leaf systemd[1]: Started Leaf API Service.
    May 08 12:11:33 omop-leaf dotnet[90214]: Unhandled exception. System.ArgumentNullException: Value cannot be null. (Parameter 'path1')
    May 08 12:11:33 omop-leaf dotnet[90214]: at System.IO.Path.Combine(String path1, String path2)
    May 08 12:11:33 omop-leaf dotnet[90214]: at API.Program.Main(String[] args) in /var/opt/leafapi/leaf_download/leaf/src/server/API/Program.cs:line 28
    May 08 12:11:33 omop-leaf systemd[1]: leaf_api.service: main process exited, code=killed, status=6/ABRT
    May 08 12:11:33 omop-leaf systemd[1]: Unit leaf_api.service entered failed state.
    May 08 12:11:33 omop-leaf systemd[1]: leaf_api.service failed.

It appears that creation of the log fails because `Environment.GetEnvironmentVariable(Config.Logging.Directory)` is null. For more details see the attached debugging data, which contains slightly annotated copies of terminal interactions.

My hypothesis is that `leaf_api.service.conf` should not be empty, but I see no instructions on what it should contain.

Regards
Arthur

[2021-05-08_debug.txt](https://github.com/uwrit/leaf/files/6446187/2021-05-08_debug.txt)

**Comment by artgoldberg** [comment]
I resolved the problem. When running in a bash shell, the environment variables need to be exported so that they're available in the dotnet process that gets launched. E.g., to make `SERILOG_DIR` available in  `Environment.GetEnvironmentVariable(Config.Logging.Directory)` it needs to be defined with:

    export SERILOG_DIR=/var/log/leaf/

**Reply by ndobb** [reply]
Thanks for letting us know, @artgoldberg. I reviewed your attached code and didn't see any issues but that makes sense. @mh2727, maybe we can incorporate some of Arthur's code into https://leafdocs.rit.uw.edu/installation/installation_steps/7_env/? We mention using a `.conf` file but aren't too explicit for Linux setup (our rationale for that if I recall correctly was that there are many ways to configure env variables and we didn't want to be overly prescriptive, though perhaps we can provide examples and be more clear about ensuring the variables are visible to the API at runtime).

**Reply by artgoldberg** [reply]
Thanks @ndobb , but it's more complicated than I thought and not fully resolved. When launching the API with `dotnet run`, the env vars are available to the C# code. That is, this code:

		Console.WriteLine($"Config.Logging.Directory: {Config.Logging.Directory}");
		Console.WriteLine($"Config.Logging.FileTemplate: {Config.Logging.FileTemplate}");
		Console.WriteLine($"Environment.GetEnvironmentVariable(Config.Logging.Directory): {Environment.GetEnvironmentVariable(Config.Logging.Directory)}");

generates this output:

    Config.Logging.Directory: SERILOG_DIR
    Config.Logging.FileTemplate: leaf-api-{Date}.log
    Environment.GetEnvironmentVariable(Config.Logging.Directory): /var/log/leaf/

But when launching the API with:

    # Make the systemd aware of the service
    systemctl daemon-reload

    # Start the service
    systemctl start leaf_api.service

the env vars are not available to the API. E. g., the same code generates this:

    [root@omop-leaf leafapi]# systemctl status -l leaf_api.service
    ● leaf_api.service - Leaf API Service
       Loaded: loaded (/var/opt/leafapi/services/leaf_api.service; linked; vendor preset: disabled)
       Active: failed (Result: signal) since Mon 2021-05-10 18:32:20 EDT; 10s ago
      Process: 99400 ExecStart=/usr/bin/dotnet API.dll (code=killed, signal=ABRT)
     Main PID: 99400 (code=killed, signal=ABRT)

    May 10 18:32:20 omop-leaf systemd[1]: Started Leaf API Service.
    May 10 18:32:20 omop-leaf dotnet[99400]: Config.Logging.Directory: SERILOG_DIR
    May 10 18:32:20 omop-leaf dotnet[99400]: Config.Logging.FileTemplate: leaf-api-{Date}.log
    May 10 18:32:20 omop-leaf dotnet[99400]: Environment.GetEnvironmentVariable(Config.Logging.Directory):
    May 10 18:32:20 omop-leaf dotnet[99400]: Unhandled exception. System.ArgumentNullException: Value cannot be null. (Parameter 'path1')
    May 10 18:32:20 omop-leaf dotnet[99400]: at System.IO.Path.Combine(String path1, String path2)
    May 10 18:32:20 omop-leaf dotnet[99400]: at API.Program.Main(String[] args) in /var/opt/leafapi/leaf_download/leaf/src/server/API/Program.cs:line 32

**Reply by artgoldberg** [reply]
Hi again @ndobb and @mh2727 . My initial tests indicate that to [use environment variables in `systemd`](https://kinvolk.io/docs/flatcar-container-linux/latest/setup/systemd/environment-variables/) on Linux we need to employ either the `Environment` command in the `leaf_api.service` file or the `EnvironmentFile` command in the `.service` file and create a `leaf_api.service.conf` file. I'll take the latter approach, so that the env vars are defined separately.

**Reply by mh2727** [reply]
Hi @artgoldberg.

The latter approach would be a cleaner way of setting up Leaf.

If it's helpful, your ```Environment``` variables should be in ```leaf_api.service.conf``` like so: https://leafdocs.rit.uw.edu/installation/installation_steps/7_env/#7-set-environment-variables.

And in your ```.service``` file, you would define the ```EnvironmentFile``` line like so:
```EnvironmentFile=/path/to/leaf_api.service.conf```.

**Reply by artgoldberg** [reply]
@ndobb and @mh2727 
Yes, that makes sense Mehadi.
Good news, the Leaf API is running.
